Design.md

```diff
@@ -136,46 +121,88 @@ This API will update the existing storage tenant with the provided `cd_tenant_id
+        * _If this role is deleted, it will be automatically repopulated (See [Assume-Role](#Assume-Role) for more)._
+1. Create an IAM managed policy `adminPolicy@[account-id]` with full S3 and IAM access using the generated access key.
+1. Invoke `attach-role-policy` to attach the policy `adminPolicy@[account-id]` to the `osis` role using the generated access key.
+1. Use `delete-access-key` to delete the account's access key.
+
 ## User APIs
 
-### Assume Role behavior for all User APIs
-* Use the tenant account credentials from the `assumeRoleCache` cache for the given account ID before each User API.
-* If the tenant account credentials were not found in `assumeRoleCache`, `AssumeRoleBackbeat` api needs to be called as Super Admin 
-* `assumeRoleCache` cache will be key-value pair of account_id and temporary credentials object respectively.
-    * `assumeRoleCache` cache will have TTL.
-    * Refresh the credentials periodically on the cache (AWS SDK `refresh()` will be used)
-    * _[ A detailed cache design needs to be added ]_
+### Common Behavior for All User APIs
+* Invoke the [Assume Role flow](#Assume-Role) before accessing any user APIs. 
+* If any user API returns an `Access Denied` error with the `user [RoleArn] don't have any policies, denied access`  message,  then invoke the `SetupAssumeRolePolicy` flow (See [Setup-Assume-Role-Policy](#Set-up-assume-role-policy).
+
+### Set Up Assume Role Policy
+1. Generate an access key for the account by using vaultclient to call `generate-account-access-key` with `expirationTime`.
+1. Invoke IAM `get-policy` API for policy `adminPolicy@[account-id]` using the generated access key.
+1. If no policy is returned, use the generated access key to create an IAM managed policy `adminPolicy@[account-id]` with full s3 and iam access (`{ s3:*, iam:* }`).
+1. Invoke `attach-role-policy` to attach the policy `adminPolicy@[account-id]` to the `osis` role using the generated access key.
+1. `delete-access-key` for the account.
 
+### Assume Role
+* Before invoking a User API, you must call the `AssumeRole` flow. Use the tenant account credentials from `assumeRoleCache` for the given account ID before each User API.
+* If the tenant account credentials were not found in `assumeRoleCache`, the `AssumeRoleBackbeat` API must be called as superadmin and added to the `assumeRoleCache` cache.
+* If the `AssumeRoleBackbeat` API returns a `Role does not exist` error, then invoke the [`SetupAssumeRole` subroutine](#SetupAssumeRole-subroutine)).
+* `assumeRoleCache` cache will be a **[key, value]** pair of **[Role_Arn, Temporary_Credentials]** respectively.
+    * Each `assumeRoleCache` entry will have a 50-minute TTL (its session token is valid only for 60 minutes).
+    * When accessing credentials on the cache, refresh the credentials as well as `needsRefresh` (using AWS SDK `refresh()`), as they are subject to expiration.
```

Comment:
I am okay with this.
Just FYI: https://scality.atlassian.net/browse/S3C-3833

We found a bug with the implementation of `needsRefresh()` in backbeat earlier this year.
Posting here, so devs are aware and we make sure that edge case is covered as well. THis will help us not miss the same edge case in OSIS. :)